File: scripts/scaffold/steps/finalize.py
"""Steps 7-12: README, SonarCloud, commit/push, verify, cleanup."""

from __future__ import annotations

import os
import shutil
import time

from ..config import Config
from ..log import fail, fatal, log, ok
from ..shell import GitHub, SonarCloud, check_rate_limit, run


def get_sonar_project_keys(cfg: Config) -> list[str]:
    prefix = f"{cfg.owner}_{cfg.repo}"
    if cfg.arch == "monolith":
        return [f"{prefix}-monolith-{cfg.lang}"]
    else:
        return [
            f"{cfg.owner}_{cfg.backend_repo}-multitier-backend-{cfg.backend_lang}",
            f"{cfg.owner}_{cfg.frontend_repo}-multitier-frontend-{cfg.frontend_lang}",
        ]


# ─── Step 7: Update README ──────────────────────────────────────────────────


def update_readme(cfg: Config, **_: object) -> None:
    log("Step 7: Generating README...")

    if cfg.dry_run:
        log("[DRY RUN] Would generate README.md")
        return

    if cfg.arch == "monolith":
        _write_readme(cfg.repo_dir, cfg.system_name, _generate_badges(cfg), cfg.owner)
    else:
        _write_system_readme(cfg)
        _write_component_readme(
            cfg.backend_repo_dir, cfg.system_name, "Backend",
            cfg.backend_full_repo, cfg.backend_lang, "backend", cfg.owner,
        )
        _write_component_readme(
            cfg.frontend_repo_dir, cfg.system_name, "Frontend",
            cfg.frontend_full_repo, cfg.frontend_lang, "frontend", cfg.owner,
        )

    ok("Generated README.md")


def _write_readme(repo_dir: str, title: str, badges: str, owner: str) -> None:
    readme = (
        f"# {title}\n\n"
        f"{badges}\n"
        f"## License\n\nMIT License\n\n"
        f"## Contributors\n\n"
        f"- [{owner}](https://github.com/{owner})\n"
    )
    with open(os.path.join(repo_dir, "README.md"), "w", encoding="utf-8", newline="\n") as f:
        f.write(readme)


def _write_system_readme(cfg: Config) -> None:
    bl, fl, tl = cfg.backend_lang, cfg.frontend_lang, cfg.test_lang
    base = f"https://github.com/{cfg.full_repo}/actions/workflows"

    # System repo badges: commit stages point to component repos
    backend_base = f"https://github.com/{cfg.backend_full_repo}/actions/workflows"
    frontend_base = f"https://github.com/{cfg.frontend_full_repo}/actions/workflows"

    badge_items = [
        (f"{backend_base}/multitier-backend-{bl}-commit-stage.yml", "backend-commit-stage"),
        (f"{frontend_base}/multitier-frontend-{fl}-commit-stage.yml", "frontend-commit-stage"),
        (f"{base}/multitier-system-{tl}-acceptance-stage.yml", "acceptance-stage"),
        (f"{base}/multitier-system-{tl}-qa-stage.yml", "qa-stage"),
        (f"{base}/multitier-system-{tl}-qa-signoff.yml", "qa-signoff"),
        (f"{base}/multitier-system-{tl}-prod-stage.yml", "prod-stage"),
    ]
    badges = "\n".join(
        f"[![{label}]({url}/badge.svg)]({url})"
        for url, label in badge_items
    ) + "\n"

    repos_section = (
        f"## Repositories\n\n"
        f"- [{cfg.backend_repo}](https://github.com/{cfg.backend_full_repo}) — Backend ({bl})\n"
        f"- [{cfg.frontend_repo}](https://github.com/{cfg.frontend_full_repo}) — Frontend ({fl})\n"
    )

    readme = (
        f"# {cfg.system_name}\n\n"
        f"{badges}\n"
        f"{repos_section}\n"
        f"## License\n\nMIT License\n\n"
        f"## Contributors\n\n"
        f"- [{cfg.owner}](https://github.com/{cfg.owner})\n"
    )
    with open(os.path.join(cfg.repo_dir, "README.md"), "w", encoding="utf-8", newline="\n") as f:
        f.write(readme)


def _write_component_readme(
    repo_dir: str, system_name: str, component_label: str,
    full_repo: str, lang: str, component_type: str, owner: str,
) -> None:
    wf_name = f"multitier-{component_type}-{lang}-commit-stage.yml"
    base = f"https://github.com/{full_repo}/actions/workflows"
    badges = f"[![commit-stage]({base}/{wf_name}/badge.svg)]({base}/{wf_name})\n"

    readme = (
        f"# {system_name} — {component_label}\n\n"
        f"{badges}\n"
        f"## License\n\nMIT License\n\n"
        f"## Contributors\n\n"
        f"- [{owner}](https://github.com/{owner})\n"
    )
    with open(os.path.join(repo_dir, "README.md"), "w", encoding="utf-8", newline="\n") as f:
        f.write(readme)


def _generate_badges(cfg: Config) -> str:
    base = f"https://github.com/{cfg.full_repo}/actions/workflows"

    if cfg.arch == "monolith":
        lang, test_lang = cfg.lang, cfg.test_lang
        items = [
            (f"monolith-{lang}-commit-stage.yml", "commit-stage"),
            (f"monolith-{test_lang}-acceptance-stage.yml", "acceptance-stage"),
            (f"monolith-{test_lang}-qa-stage.yml", "qa-stage"),
            (f"monolith-{test_lang}-qa-signoff.yml", "qa-signoff"),
            (f"monolith-{test_lang}-prod-stage.yml", "prod-stage"),
        ]
    else:
        bl, fl, tl = cfg.backend_lang, cfg.frontend_lang, cfg.test_lang
        items = [
            (f"multitier-backend-{bl}-commit-stage.yml", "backend-commit-stage"),
            (f"multitier-frontend-{fl}-commit-stage.yml", "frontend-commit-stage"),
            (f"multitier-system-{tl}-acceptance-stage.yml", "acceptance-stage"),
            (f"multitier-system-{tl}-qa-stage.yml", "qa-stage"),
            (f"multitier-system-{tl}-qa-signoff.yml", "qa-signoff"),
            (f"multitier-system-{tl}-prod-stage.yml", "prod-stage"),
        ]

    return "\n".join(
        f"[![{label}]({base}/{wf}/badge.svg)]({base}/{wf})"
        for wf, label in items
    ) + "\n"


# ─── Step 8: Create SonarCloud projects ─────────────────────────────────────


def create_sonarcloud_projects(cfg: Config, sonarcloud: SonarCloud, **_: object) -> None:
    log("Step 8: Creating SonarCloud projects...")

    if cfg.dry_run:
        log("[DRY RUN] Would create SonarCloud org and project(s)")
        return

    sonarcloud.create_org()
    for key in get_sonar_project_keys(cfg):
        sonarcloud.create_project(key)


# ─── Step 9: Commit and push ────────────────────────────────────────────────


def commit_and_push(cfg: Config, **_: object) -> None:
    log("Step 9: Committing and pushing...")

    if cfg.dry_run:
        log("[DRY RUN] Would git add, commit, push")
        return

    _commit_and_push_repo(cfg.repo_dir, cfg.full_repo)

    if cfg.arch == "multitier":
        _commit_and_push_repo(cfg.backend_repo_dir, cfg.backend_full_repo)
        _commit_and_push_repo(cfg.frontend_repo_dir, cfg.frontend_full_repo)


def _commit_and_push_repo(repo_dir: str, full_repo: str) -> None:
    run("git add -A", cwd=repo_dir)
    run('git commit -m "Apply pipeline template"', cwd=repo_dir)
    run("git push", cwd=repo_dir)
    ok(f"Pushed template to {full_repo}")


# ─── Step 10: Verify commit stage ───────────────────────────────────────────


def verify_commit_stage(cfg: Config, github: GitHub, **_: object) -> None:
    log("Step 10: Verifying commit stage workflow...")

    if cfg.dry_run:
        log("[DRY RUN] Would wait for commit stage workflow")
        return

    time.sleep(5)

    if cfg.arch == "monolith":
        _verify_workflow(github, "Commit stage")
    else:
        # Wait for commit stages on both component repos
        gh_backend = github.for_repo(cfg.backend_full_repo)
        gh_frontend = github.for_repo(cfg.frontend_full_repo)

        _verify_workflow(gh_backend, "Backend commit stage")
        _verify_workflow(gh_frontend, "Frontend commit stage")


# ─── Step 11: Verify acceptance stage ────────────────────────────────────────


def verify_acceptance_stage(cfg: Config, github: GitHub, **_: object) -> None:
    log("Step 11: Triggering and verifying acceptance stage...")

    if cfg.dry_run:
        log("[DRY RUN] Would trigger and wait for acceptance stage workflow")
        return

    test_lang = cfg.test_lang
    if cfg.arch == "monolith":
        wf = f"monolith-{test_lang}-acceptance-stage.yml"
    else:
        wf = f"multitier-system-{test_lang}-acceptance-stage.yml"

    _verify_workflow(github, "Acceptance stage", trigger_workflow=wf)

    # Extract RC version from the acceptance stage run for downstream stages
    rc_version = _get_rc_version(github)
    if rc_version:
        cfg.rc_version = rc_version  # type: ignore[attr-defined]
        ok(f"RC version: {rc_version}")
    else:
        fatal("Could not determine RC version from acceptance stage run")


# ─── Step 12: Verify QA stage ───────────────────────────────────────────────


def verify_qa_stage(cfg: Config, github: GitHub, **_: object) -> None:
    log("Step 12: Triggering and verifying QA stage...")

    if cfg.dry_run:
        log("[DRY RUN] Would trigger and wait for QA stage workflow")
        return

    rc_version = getattr(cfg, "rc_version", None)
    if not rc_version:
        fatal("No RC version available — acceptance stage must run first")

    test_lang = cfg.test_lang
    if cfg.arch == "monolith":
        wf = f"monolith-{test_lang}-qa-stage.yml"
    else:
        wf = f"multitier-system-{test_lang}-qa-stage.yml"

    _verify_workflow(github, "QA stage", trigger_workflow=wf,
                     fields={"version": rc_version})


# ─── Step 13: Verify QA signoff ─────────────────────────────────────────────


def verify_qa_signoff(cfg: Config, github: GitHub, **_: object) -> None:
    log("Step 13: Triggering and verifying QA signoff...")

    if cfg.dry_run:
        log("[DRY RUN] Would trigger and wait for QA signoff workflow")
        return

    rc_version = getattr(cfg, "rc_version", None)
    if not rc_version:
        fatal("No RC version available — acceptance stage must run first")

    test_lang = cfg.test_lang
    if cfg.arch == "monolith":
        wf = f"monolith-{test_lang}-qa-signoff.yml"
    else:
        wf = f"multitier-system-{test_lang}-qa-signoff.yml"

    _verify_workflow(github, "QA signoff", trigger_workflow=wf,
                     fields={"version": rc_version, "result": "approved"})


# ─── Step 14: Verify production stage ───────────────────────────────────────


def verify_prod_stage(cfg: Config, github: GitHub, **_: object) -> None:
    log("Step 14: Triggering and verifying production stage...")

    if cfg.dry_run:
        log("[DRY RUN] Would trigger and wait for production stage workflow")
        return

    rc_version = getattr(cfg, "rc_version", None)
    if not rc_version:
        fatal("No RC version available — acceptance stage must run first")

    test_lang = cfg.test_lang
    if cfg.arch == "monolith":
        wf = f"monolith-{test_lang}-prod-stage.yml"
    else:
        wf = f"multitier-system-{test_lang}-prod-stage.yml"

    _verify_workflow(github, "Production stage", trigger_workflow=wf,
                     fields={"version": rc_version})


# ─── Workflow helpers ────────────────────────────────────────────────────────


def _verify_workflow(github: GitHub, label: str, trigger_workflow: str | None = None,
                     fields: dict[str, str] | None = None) -> None:
    check_rate_limit()
    if trigger_workflow:
        github.workflow_run(trigger_workflow, fields=fields)
        time.sleep(5)

    check_rate_limit()
    result = github.run_watch()
    if result.returncode != 0:
        fail(f"{label} failed!")
        fatal(f"{label} workflow failed. Check: https://github.com/{github.repo}/actions")
    ok(f"{label} passed!")


def _get_rc_version(github: GitHub) -> str | None:
    """Get the latest RC version from GitHub releases."""
    import json as _json

    check_rate_limit()
    result = run(
        f"gh api repos/{github.repo}/releases --jq .[0].tag_name",
        check=False, capture=True,
    )
    tag = result.stdout.strip()
    if tag and "-rc." in tag:
        return tag

    # Fallback: list releases as JSON and parse
    result = run(
        f"gh api repos/{github.repo}/releases",
        check=False, capture=True,
    )
    try:
        releases = _json.loads(result.stdout)
        if releases:
            tag = releases[0].get("tag_name", "")
            if "-rc." in tag:
                return tag
    except (ValueError, IndexError, KeyError):
        pass

    return None

File: gh-optivem/internal/steps/finalize.go
package steps

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/optivem/gh-optivem/internal/config"
	"github.com/optivem/gh-optivem/internal/log"
	"github.com/optivem/gh-optivem/internal/shell"
)

// GetSonarProjectKeys returns the SonarCloud project keys for the given config.
func GetSonarProjectKeys(cfg *config.Config) []string {
	prefix := cfg.Owner + "_" + cfg.Repo
	if cfg.Arch == "monolith" {
		return []string{prefix + "-monolith-" + cfg.Lang}
	}
	return []string{
		cfg.Owner + "_" + cfg.BackendRepo + "-multitier-backend-" + cfg.BackendLang,
		cfg.Owner + "_" + cfg.FrontendRepo + "-multitier-frontend-" + cfg.FrontendLang,
	}
}

// UpdateReadme generates README.md for the repo(s).
func UpdateReadme(cfg *config.Config) {
	log.Log("Step 8: Generating README...")

	if cfg.DryRun {
		log.Log("[DRY RUN] Would generate README.md")
		return
	}

	if cfg.Arch == "monolith" {
		badges := generateBadges(cfg)
		writeReadme(cfg.RepoDir, cfg.SystemName, badges, cfg.Owner)
	} else {
		writeSystemReadme(cfg)
		writeComponentReadme(
			cfg.BackendRepoDir, cfg.SystemName, "Backend",
			cfg.BackendFullRepo, cfg.BackendLang, "backend", cfg.Owner,
		)
		writeComponentReadme(
			cfg.FrontendRepoDir, cfg.SystemName, "Frontend",
			cfg.FrontendFullRepo, cfg.FrontendLang, "frontend", cfg.Owner,
		)
	}

	log.OK("Generated README.md")
}

func writeReadme(repoDir, title, badges, owner string) {
	content := fmt.Sprintf("# %s\n\n%s\n## License\n\nMIT License\n\n## Contributors\n\n- [%s](https://github.com/%s)\n",
		title, badges, owner, owner)
	os.WriteFile(filepath.Join(repoDir, "README.md"), []byte(content), 0644)
}

func writeSystemReadme(cfg *config.Config) {
	bl, fl, tl := cfg.BackendLang, cfg.FrontendLang, cfg.TestLang
	base := "https://github.com/" + cfg.FullRepo + "/actions/workflows"
	backendBase := "https://github.com/" + cfg.BackendFullRepo + "/actions/workflows"
	frontendBase := "https://github.com/" + cfg.FrontendFullRepo + "/actions/workflows"

	badgeItems := [][2]string{
		{backendBase + "/multitier-backend-" + bl + "-commit-stage.yml", "backend-commit-stage"},
		{frontendBase + "/multitier-frontend-" + fl + "-commit-stage.yml", "frontend-commit-stage"},
		{base + "/multitier-system-" + tl + "-acceptance-stage.yml", "acceptance-stage"},
		{base + "/multitier-system-" + tl + "-qa-stage.yml", "qa-stage"},
		{base + "/multitier-system-" + tl + "-qa-signoff.yml", "qa-signoff"},
		{base + "/multitier-system-" + tl + "-prod-stage.yml", "prod-stage"},
	}

	var badges strings.Builder
	for _, item := range badgeItems {
		fmt.Fprintf(&badges, "[![%s](%s/badge.svg)](%s)\n", item[1], item[0], item[0])
	}

	reposSection := fmt.Sprintf("## Repositories\n\n- [%s](https://github.com/%s) — Backend (%s)\n- [%s](https://github.com/%s) — Frontend (%s)\n",
		cfg.BackendRepo, cfg.BackendFullRepo, bl,
		cfg.FrontendRepo, cfg.FrontendFullRepo, fl)

	content := fmt.Sprintf("# %s\n\n%s\n%s\n## License\n\nMIT License\n\n## Contributors\n\n- [%s](https://github.com/%s)\n",
		cfg.SystemName, badges.String(), reposSection, cfg.Owner, cfg.Owner)
	os.WriteFile(filepath.Join(cfg.RepoDir, "README.md"), []byte(content), 0644)
}

func writeComponentReadme(repoDir, systemName, componentLabel, fullRepo, lang, componentType, owner string) {
	wfName := "multitier-" + componentType + "-" + lang + "-commit-stage.yml"
	base := "https://github.com/" + fullRepo + "/actions/workflows"
	badges := fmt.Sprintf("[![commit-stage](%s/%s/badge.svg)](%s/%s)\n", base, wfName, base, wfName)

	content := fmt.Sprintf("# %s — %s\n\n%s\n## License\n\nMIT License\n\n## Contributors\n\n- [%s](https://github.com/%s)\n",
		systemName, componentLabel, badges, owner, owner)
	os.WriteFile(filepath.Join(repoDir, "README.md"), []byte(content), 0644)
}

func generateBadges(cfg *config.Config) string {
	base := "https://github.com/" + cfg.FullRepo + "/actions/workflows"

	var items [][2]string
	if cfg.Arch == "monolith" {
		items = [][2]string{
			{"monolith-" + cfg.Lang + "-commit-stage.yml", "commit-stage"},
			{"monolith-" + cfg.TestLang + "-acceptance-stage.yml", "acceptance-stage"},
			{"monolith-" + cfg.TestLang + "-qa-stage.yml", "qa-stage"},
			{"monolith-" + cfg.TestLang + "-qa-signoff.yml", "qa-signoff"},
			{"monolith-" + cfg.TestLang + "-prod-stage.yml", "prod-stage"},
		}
	} else {
		bl, fl, tl := cfg.BackendLang, cfg.FrontendLang, cfg.TestLang
		items = [][2]string{
			{"multitier-backend-" + bl + "-commit-stage.yml", "backend-commit-stage"},
			{"multitier-frontend-" + fl + "-commit-stage.yml", "frontend-commit-stage"},
			{"multitier-system-" + tl + "-acceptance-stage.yml", "acceptance-stage"},
			{"multitier-system-" + tl + "-qa-stage.yml", "qa-stage"},
			{"multitier-system-" + tl + "-qa-signoff.yml", "qa-signoff"},
			{"multitier-system-" + tl + "-prod-stage.yml", "prod-stage"},
		}
	}

	var b strings.Builder
	for _, item := range items {
		fmt.Fprintf(&b, "[![%s](%s/%s/badge.svg)](%s/%s)\n", item[1], base, item[0], base, item[0])
	}
	return b.String()
}

// CreateSonarCloudProjects creates SonarCloud org and projects.
func CreateSonarCloudProjects(cfg *config.Config, sc *shell.SonarCloud) {
	log.Log("Step 9: Creating SonarCloud projects...")

	if cfg.DryRun {
		log.Log("[DRY RUN] Would create SonarCloud org and project(s)")
		return
	}

	sc.CreateOrg()
	for _, key := range GetSonarProjectKeys(cfg) {
		sc.CreateProject(key)
	}
}

// CommitAndPush commits and pushes changes to GitHub.
func CommitAndPush(cfg *config.Config) {
	log.Log("Step 10: Committing and pushing...")

	if cfg.DryRun {
		log.Log("[DRY RUN] Would git add, commit, push")
		return
	}

	commitAndPushRepo(cfg.RepoDir, cfg.FullRepo)

	if cfg.Arch == "multitier" {
		commitAndPushRepo(cfg.BackendRepoDir, cfg.BackendFullRepo)
		commitAndPushRepo(cfg.FrontendRepoDir, cfg.FrontendFullRepo)
	}
}

func commitAndPushRepo(repoDir, fullRepo string) {
	shell.Run("git add -A", false, true, repoDir)
	shell.Run(`git commit -m "Apply pipeline template"`, false, true, repoDir)
	shell.Run("git push", false, true, repoDir)
	log.OKf("Pushed template to %s", fullRepo)
}

// VerifyCommitStage waits for commit stage workflow to pass.
func VerifyCommitStage(cfg *config.Config, gh *shell.GitHub) {
	log.Log("Step 11: Verifying commit stage workflow...")

	if cfg.DryRun {
		log.Log("[DRY RUN] Would wait for commit stage workflow")
		return
	}

	time.Sleep(5 * time.Second)

	if cfg.Arch == "monolith" {
		verifyWorkflow(gh, "Commit stage", "", nil)
	} else {
		ghBackend := gh.ForRepo(cfg.BackendFullRepo)
		ghFrontend := gh.ForRepo(cfg.FrontendFullRepo)
		verifyWorkflow(ghBackend, "Backend commit stage", "", nil)
		verifyWorkflow(ghFrontend, "Frontend commit stage", "", nil)
	}
}

// VerifyAcceptanceStage triggers and verifies acceptance stage.
func VerifyAcceptanceStage(cfg *config.Config, gh *shell.GitHub) {
	log.Log("Step 12: Triggering and verifying acceptance stage...")

	if cfg.DryRun {
		log.Log("[DRY RUN] Would trigger and wait for acceptance stage workflow")
		return
	}

	var wf string
	if cfg.Arch == "monolith" {
		wf = "monolith-" + cfg.TestLang + "-acceptance-stage.yml"
	} else {
		wf = "multitier-system-" + cfg.TestLang + "-acceptance-stage.yml"
	}

	verifyWorkflow(gh, "Acceptance stage", wf, nil)

	rcVersion := getRCVersion(gh)
	if rcVersion != "" {
		cfg.RCVersion = rcVersion
		log.OKf("RC version: %s", rcVersion)
	} else {
		log.Fatal("Could not determine RC version from acceptance stage run")
	}
}

// VerifyQAStage triggers and verifies QA stage.
func VerifyQAStage(cfg *config.Config, gh *shell.GitHub) {
	log.Log("Step 13: Triggering and verifying QA stage...")

	if cfg.DryRun {
		log.Log("[DRY RUN] Would trigger and wait for QA stage workflow")
		return
	}

	if cfg.RCVersion == "" {
		log.Fatal("No RC version available — acceptance stage must run first")
	}

	var wf string
	if cfg.Arch == "monolith" {
		wf = "monolith-" + cfg.TestLang + "-qa-stage.yml"
	} else {
		wf = "multitier-system-" + cfg.TestLang + "-qa-stage.yml"
	}

	verifyWorkflow(gh, "QA stage", wf, map[string]string{"version": cfg.RCVersion})
}

// VerifyQASignoff triggers and verifies QA signoff.
func VerifyQASignoff(cfg *config.Config, gh *shell.GitHub) {
	log.Log("Step 14: Triggering and verifying QA signoff...")

	if cfg.DryRun {
		log.Log("[DRY RUN] Would trigger and wait for QA signoff workflow")
		return
	}

	if cfg.RCVersion == "" {
		log.Fatal("No RC version available — acceptance stage must run first")
	}

	var wf string
	if cfg.Arch == "monolith" {
		wf = "monolith-" + cfg.TestLang + "-qa-signoff.yml"
	} else {
		wf = "multitier-system-" + cfg.TestLang + "-qa-signoff.yml"
	}

	verifyWorkflow(gh, "QA signoff", wf, map[string]string{"version": cfg.RCVersion, "result": "approved"})
}

// VerifyProdStage triggers and verifies production stage.
func VerifyProdStage(cfg *config.Config, gh *shell.GitHub) {
	log.Log("Step 15: Triggering and verifying production stage...")

	if cfg.DryRun {
		log.Log("[DRY RUN] Would trigger and wait for production stage workflow")
		return
	}

	if cfg.RCVersion == "" {
		log.Fatal("No RC version available — acceptance stage must run first")
	}

	var wf string
	if cfg.Arch == "monolith" {
		wf = "monolith-" + cfg.TestLang + "-prod-stage.yml"
	} else {
		wf = "multitier-system-" + cfg.TestLang + "-prod-stage.yml"
	}

	verifyWorkflow(gh, "Production stage", wf, map[string]string{"version": cfg.RCVersion})
}

func verifyWorkflow(gh *shell.GitHub, label, triggerWorkflow string, fields map[string]string) {
	shell.CheckRateLimit()
	if triggerWorkflow != "" {
		gh.WorkflowRun(triggerWorkflow, fields)
		time.Sleep(5 * time.Second)
	}

	shell.CheckRateLimit()
	err := gh.RunWatch()
	if err != nil {
		log.Failf("%s failed!", label)
		log.Fatalf("%s workflow failed. Check: https://github.com/%s/actions", label, gh.Repo)
	}
	log.OKf("%s passed!", label)
}

func getRCVersion(gh *shell.GitHub) string {
	shell.CheckRateLimit()

	out, err := shell.RunCapture(
		fmt.Sprintf("gh api repos/%s/releases --jq .[0].tag_name", gh.Repo), "")
	if err == nil && strings.Contains(out, "-rc.") {
		return out
	}

	// Fallback: parse JSON
	out, err = shell.RunCapture(
		fmt.Sprintf("gh api repos/%s/releases", gh.Repo), "")
	if err == nil {
		var releases []struct {
			TagName string `json:"tag_name"`
		}
		if json.Unmarshal([]byte(out), &releases) == nil && len(releases) > 0 {
			if strings.Contains(releases[0].TagName, "-rc.") {
				return releases[0].TagName
			}
		}
	}

	return ""
}
